feat(metrics): add datadog statsd to metrics

aelath opened the pull request:
Simple request metrics are now also sent to the datadog statsd agent.

Commit:
feat(metrics): add datadog statsd to metrics

aelath commented on middlewares/metrics.go:
```diff
@@ -1,12 +1,35 @@
+	applicationNamespace := os.Getenv("STATSD_APP_NAME")
+
+	var err error
+	statsdClient, err = statsd.New(statsdHost + ":" + statsdPort)
```
I've gone back and forth on defaulting these. I'd like to fail fast, but there is a chance someone will be using this without dogstatsd.

bnewtonius replied:
I didn't fail fast in the node implementations of these. I've also been bitten by not failing fast before. For this though, I'm ok with letting the default of "localhost:8125" be the fallback. We may change our minds later and that's ok too. We don't currently have a running statsd agent anyways, so we'd also need to have an on/off switch if we plan on failing fast.

bnewtonius commented on middlewares/metrics.go:
```diff
@@ -1,12 +1,35 @@
+	var err error
+	statsdClient, err = statsd.New(statsdHost + ":" + statsdPort)
+	if err != nil {
+		fmt.Println("** We were unable to get a statsd client.")
```
I typically do not like to have err checks like this in a library. It is better to return the error to the caller and let them decide what do do. That way they can do logging properly rather than just printing to stdout, fail fast, fallback, turn off, whatever.

bnewtonius commented on middlewares/metrics.go:
```diff
@@ -1,12 +1,35 @@
+var statsdClient *statsd.Client
+
+func initStatsdClient() {
+	statsdHost := os.Getenv("STATSD_HOST")
```
If you wanted to make this lib more flexible, accept these as arguments to the function. That way the caller can get them as either env vars or flags or whatever.

bnewtonius commented on middlewares/metrics.go:
```diff
@@ -1,12 +1,35 @@
+	statsdClient.Namespace = applicationNamespace
+}
+
+func getStatsdClient() *statsd.Client {
```
Minor: golang recommends not calling getters "getSomething", but rather just calling it "something". So this would be called `statsdClient()`, or even just `client()` if you were to move this to its own namespace called "statsd"

bnewtonius reviewed (commented):
In general, you may want to move the statsd stuff into its own namespace/package.

Commit:
respond to brians feedback

aelath commented:
@bnewtonius I updated this PR. I moved the statsd off into its own file, although it is still part of the middlewares pkg. In the consumer (in this case warden) I call `InitClient` in the init function of serve.go. I'm a bit iffy on how to separate out the concerns inside the middlewares pkg; any thoughts are much appreciated.
